Convert channel links

scgruber opened the pull request:
This PR converts InputChannelLink and OutputChannelLink to the new LinkBuilder style. It also fixes the examples to be compliant with the new style. I'm not updating graphgen in this PR because this already feels substaintial.

Marked as Draft for now because I haven't finished OutputChannelLink.

scgruber committed:
Fix examples to work with new LinkBuilder style

scgruber committed:
InputChannelLink: Convert to Link interface

This one is a little weird in that it does not take ingress streams,
unlike all of the other links. Thus, we panic if ingressors is ever
called. Otherwise this behaves pretty similarly to ProcessLink.

A subsequent commit will convert OutputChannelLink in a similar way.

ddowl commented on examples/trivial-identity/src/pipeline.rs:
```diff
@@ -17,15 +18,25 @@ impl route_rs_runtime::pipeline::Runner for Pipeline {
-        let link_3 = OutputChannelLink::new(Box::new(link_2), output_channel);
+        let link_3 = OutputChannelLink::new(link_2_egress_0, output_channel);
 
         tokio::run(lazy(move || {
```
We should have a common function for "run all tasks on tokio".

scgruber replied:
Yeah I think I'll add this as a separate PR when I'm working on graphgen

ddowl commented on examples/dns-interceptor/src/pipeline.rs:
```diff
@@ -17,43 +18,53 @@ impl route_rs_runtime::pipeline::Runner for Pipeline {
         let elem_2_classifydns = ClassifyDNS::new();
         let elem_3_localdnsinterceptor = LocalDNSInterceptor::new();
 
         let link_1 = InputChannelLink::new(input_channel);
```
It would be nice if the links had more descriptive names, but if this is a code generation limitation then it's fine.

scgruber replied:
I think it's sort of a question of how we expect people to be working with the pipeline. Right now the Links are kind of invisible, and what you see is the Processors and such. I do think we could do more things to make the file navigable in big projects.

ddowl commented on examples/dns-interceptor/src/pipeline.rs:
```diff
@@ -17,43 +18,53 @@ impl route_rs_runtime::pipeline::Runner for Pipeline {
         let link_1 = InputChannelLink::new(input_channel);
 
-        let link_2 = ProcessLink::new(Box::new(link_1), elem_1_setinterfacebydestination);
+        let (mut runnables_2, mut link_2) = ProcessLink::new()
```
The second tuple element is the egressors of the built `Link`.

ddowl commented on route-rs-runtime/src/link/primitive/input_channel_link.rs:
```diff
@@ -0,0 +1,98 @@
+        let link = InputChannelLink::new().channel(recv).build_link();
+
+        for p in packets.clone() {
+            match send.send(p) {
```
lol `send.send()`. Maybe we can call the `Sender` and `Receiver` `(s, r)`? That's what I've seen in other crossbeam examples.

scgruber committed:
OutputChannelLink: convert to Link interface

Similar to the previous commit, let's standardize the Links on the new
interface. OutputChannelLink does take an ingressor but it produces no
egressors since everything exits through the channel.

This commit also updates the examples to use the new link style.

CollinValley commented on route-rs-runtime/src/link/primitive/input_channel_link.rs:
```diff
@@ -0,0 +1,98 @@
+        } else {
+            (
+                vec![],
+                vec![Box::new(StreamFromChannel {
```
This is pretty nifty.  Makes sense that this would only be an egressor, and has no tokio-runnable cost.

CollinValley commented on route-rs-runtime/src/link/primitive/output_channel_link.rs:
```diff
@@ -0,0 +1,170 @@
+            if self.channel_sender.is_full() {
+                // Since we don't know anything about the other side of our channel, we have to
+                // self-notify and just hope that the other side empties it eventually.
+                task::current().notify();
```
This is expensive to do. It is 100% better to do than a tight spin loop, and I understand why we do it. But this means we should only making a habit of using the egressor-channel interface sparingly.

Ideally we would have some way to subscribe ourselves to be awoken by something in this interface, something to watch out for when we go try to plug this in to whatever is sending the packets to the interface.

CollinValley reviewed (dismissed):
Overall this looks good to me. 

I think it is important we don't use this a lot cause it is very greedy with the self-notify. But it is fit for purpose.

You may be able to use this to replace all our dedicated collectors in the testing harness. It is essentially the same ish.

CollinValley reviewed (dismissed):
LGTMMD

ddowl reviewed (approved):
LGTMMMFD
